File: rust/tests/link_tests.rs
use ln::Ln;

#[test]
fn link_constructor_with_id_only_test() {
    let link = Ln::Link::<String> {
        id: Some("test".to_string()),
        values: vec![],
    };
    if let Ln::Link { id, values } = link {
        assert_eq!(id, Some("test".to_string()));
        assert!(values.is_empty());
    } else {
        panic!("Expected Link variant");
    }
}

#[test]
fn link_constructor_with_id_and_values_test() {
    let values = vec![
        Ln::Ref("value1".to_string()),
        Ln::Ref("value2".to_string()),
    ];
    let link = Ln::Link {
        id: Some("parent".to_string()),
        values: values.clone(),
    };
    if let Ln::Link { id, values: link_values } = link {
        assert_eq!(id, Some("parent".to_string()));
        assert_eq!(link_values.len(), 2);
    } else {
        panic!("Expected Link variant");
    }
}

#[test]
fn link_to_string_with_id_only_test() {
    let link = Ln::Link::<String> {
        id: Some("test".to_string()),
        values: vec![],
    };
    assert_eq!(link.to_string(), "(test: )");
}

#[test]
fn link_to_string_with_values_only_test() {
    let values = vec![
        Ln::Ref("value1".to_string()),
        Ln::Ref("value2".to_string()),
    ];
    let link = Ln::Link::<String> {
        id: None,
        values,
    };
    assert_eq!(link.to_string(), "(value1 value2)");
}

#[test]
fn link_to_string_with_id_and_values_test() {
    let values = vec![
        Ln::Ref("child1".to_string()),
        Ln::Ref("child2".to_string()),
    ];
    let link = Ln::Link {
        id: Some("parent".to_string()),
        values,
    };
    assert_eq!(link.to_string(), "(parent: child1 child2)");
}

#[test]
fn link_equals_test() {
    let link1 = Ln::Link::<String> {
        id: Some("test".to_string()),
        values: vec![],
    };
    let link2 = Ln::Link::<String> {
        id: Some("test".to_string()),
        values: vec![],
    };
    let link3 = Ln::Link::<String> {
        id: Some("other".to_string()),
        values: vec![],
    };
    
    assert_eq!(link1, link2);
    assert_ne!(link1, link3);
}

#[test]
fn link_combine_test() {
    // Rust doesn't have a direct combine method, but we can test creating combined structures
    let link1 = Ln::Ref("first".to_string());
    let link2 = Ln::Ref("second".to_string());
    let combined = Ln::Link::<String> {
        id: None,
        values: vec![link1, link2],
    };
    
    if let Ln::Link { id, values } = combined {
        assert_eq!(id, None);
        assert_eq!(values.len(), 2);
    } else {
        panic!("Expected Link variant");
    }
}

#[test]
fn link_escape_reference_simple_test() {
    let ref_simple = Ln::Ref("test".to_string());
    assert_eq!(ref_simple.to_string(), "test");
    
    let ref_with_nums = Ln::Ref("test123".to_string());
    assert_eq!(ref_with_nums.to_string(), "test123");
}

#[test]
fn link_escape_reference_with_special_characters_test() {
    // In Rust implementation, special character handling would be in the parser
    // These tests verify basic string handling
    let ref_space = Ln::Ref("has space".to_string());
    assert_eq!(ref_space.to_string(), "has space");
    
    let ref_colon = Ln::Ref("has:colon".to_string());
    assert_eq!(ref_colon.to_string(), "has:colon");
}

#[test]
fn link_simplify_test() {
    // Test simplification behavior - empty values
    let link1 = Ln::Link::<String> {
        id: Some("test".to_string()),
        values: vec![],
    };
    // In Rust, we don't have a simplify method, but we can test the structure
    if let Ln::Link { id, values } = link1 {
        assert_eq!(id, Some("test".to_string()));
        assert!(values.is_empty());
    }

    // Test with single value
    let single_ref = Ln::Ref("single".to_string());
    let link2 = Ln::Link::<String> {
        id: None,
        values: vec![single_ref.clone()],
    };
    if let Ln::Link { id, values } = link2 {
        assert_eq!(id, None);
        assert_eq!(values.len(), 1);
        assert_eq!(values[0], single_ref);
    }
}

File: csharp/LinkFoundation.LinksNotation.Tests/LinkTests.cs
using System;
using Xunit;
using System.Collections.Generic;

namespace Platform.Protocols.Lino.Tests
{
    public static class LinkTests
    {
        [Fact]
        public static void LinkConstructorWithIdOnlyTest()
        {
            var link = new Link<string>("test", null);
            Assert.Equal("test", link.Id);
            Assert.Null(link.Values);
        }

        [Fact]
        public static void LinkConstructorWithIdAndValuesTest()
        {
            var values = new List<Link<string>> { new Link<string>("value1"), new Link<string>("value2") };
            var link = new Link<string>("parent", values);
            Assert.Equal("parent", link.Id);
            Assert.Equal(2, link.Values?.Count);
        }

        [Fact]
        public static void LinkToStringWithIdOnlyTest()
        {
            var link = new Link<string>("test");
            Assert.Equal("(test)", link.ToString());
        }

        [Fact]
        public static void LinkToStringWithValuesOnlyTest()
        {
            var values = new List<Link<string>> { new Link<string>("value1"), new Link<string>("value2") };
            var link = new Link<string>(null, values);
            Assert.Equal("(value1 value2)", link.ToString());
        }

        [Fact]
        public static void LinkToStringWithIdAndValuesTest()
        {
            var values = new List<Link<string>> { new Link<string>("child1"), new Link<string>("child2") };
            var link = new Link<string>("parent", values);
            Assert.Equal("(parent: child1 child2)", link.ToString());
        }

        [Fact]
        public static void LinkEqualsTest()
        {
            var link1 = new Link<string>("test");
            var link2 = new Link<string>("test");
            var link3 = new Link<string>("other");
            
            Assert.Equal(link1, link2);
            Assert.NotEqual(link1, link3);
        }

        [Fact]
        public static void LinkCombineTest()
        {
            var link1 = new Link<string>("first");
            var link2 = new Link<string>("second");
            var combined = link1.Combine(link2);
            
            Assert.Null(combined.Id);
            Assert.Equal(2, combined.Values?.Count);
            Assert.Equal("first", combined.Values?[0].Id);
            Assert.Equal("second", combined.Values?[1].Id);
        }

        [Fact]
        public static void LinkEscapeReferenceSimpleTest()
        {
            Assert.Equal("test", Link<string>.EscapeReference("test"));
            Assert.Equal("test123", Link<string>.EscapeReference("test123"));
        }

        [Fact]
        public static void LinkEscapeReferenceWithSpecialCharactersTest()
        {
            Assert.Equal("'has space'", Link<string>.EscapeReference("has space"));
            Assert.Equal("'has:colon'", Link<string>.EscapeReference("has:colon"));
            Assert.Equal("'has(paren)'", Link<string>.EscapeReference("has(paren)"));
            Assert.Equal("'has)paren'", Link<string>.EscapeReference("has)paren"));
            Assert.Equal("\"has'quote\"", Link<string>.EscapeReference("has'quote"));
            Assert.Equal("'has\"quote'", Link<string>.EscapeReference("has\"quote"));
        }

        [Fact]
        public static void LinkSimplifyTest()
        {
            // Test simplify with empty values
            var link1 = new Link<string>("test", null);
            var simplified1 = link1.Simplify();
            Assert.Equal(link1, simplified1);

            // Test simplify with single value
            var link2 = new Link<string>(null, new List<Link<string>> { new Link<string>("single", null) });
            var simplified2 = link2.Simplify();
            Assert.Equal("single", simplified2.Id);
            Assert.Null(simplified2.Values);
        }
    }
}